tests/test_globalconstraints.py

```diff
@@ -169,6 +169,70 @@ def test_not_circuit(self):
 
         self.assertEqual(total, len(circuit_sols) + len(not_circuit_sols))
 
+    def test_subcircuit(self):
+
+        def model_and_solve(n, not_included, startIndex=None, negative=False, decompose=False):
+            x = cp.intvar(0, n-1, n)
+            c = []
+            if not negative:
+                # SubCircuit constraint
+                if (not decompose):
+                    c += [cp.SubCircuit(x, startIndex=startIndex)]
+                else:
+                    constraining, defining = cp.SubCircuit(x, startIndex=startIndex).decompose()
+                    c += constraining
+                    c += defining
+                # Side constraints based on test case restrictions
+                c += [x[i] == i for i in not_included]
+                c += [x[i] != i for i in range(n) if i not in not_included]
+            else:
+                c += [~cp.SubCircuit(x, startIndex=startIndex)]
+                # no decompose in the negative case
+                # c += ~cp.all(list(cp.SubCircuit(x, startIndex=startIndex).decompose()))
+
+            model = cp.Model(c)
+            sat = model.solve()
+            if sat:
+                return sat, x.value()
+            return sat, None
+        
+        def included(x):
+            return [xi for i, xi in enumerate(x) if xi != i]
+
+        def assert_positive(n, not_included, startIndex, decompose):
+            sat, x = model_and_solve(n, not_included, startIndex, negative=False, decompose=decompose)
+            self.assertTrue(sat)
+            self.assertEqual(len(included(x)), (n - len(not_included)))
+            self.assertTrue(cp.SubCircuit(x))
+            for i in not_included:
+                assert(i not in included(x))
+            if startIndex is not None:
+                assert(startIndex in included(x))
+
+        def assert_negative(n, not_included, startIndex, decompose):
```

Comment:
If you use the Choco solver here, it will probably fail

Reply:
No, since the negation of the subcircuit (the pure choco one, without the constraint for the start_index which for now is wrongly added at top level) does not contradict the start_index constraint (that the start_index stop cannot self-loop) 

For example [1, 0, 3, 2] satisfies choco's ~subcircuit, since two subcircuits are not allowed, and satisfies the start_index constraint if 0 was chosen as start_index.

What I think you hint at is the following example:
```python
n = 6
start_index = 0
x = cp.intvar(0, n-1, n)

c = []

a = cp.boolvar()
c += [~cp.SubCircuit(x, start_index=start_index)]
c += [x[start_index] == start_index]

model = cp.Model(c)
s = cp.SolverLookup.get("choco", model)
assert(s.solve())

print(x.value())
```
Here I enforce that the start_index must self-loop, which contradicts the start_index condition. In a correct implementation, this should not cause any issues since the start_index constraint is inside subcircuit which appears in a negated setting. But currently the start_index condition gets posted at top level and as you suspected the problem becomes UNSAT.